File: bindings/python/satox/core/blockchain_manager.py
"""
Blockchain Manager Python Bindings
"""

from typing import Dict, List, Optional, Any, Tuple
import satox_cpp
from satox_bindings.core.thread_safe_manager import ThreadSafeManager, ThreadState

class BlockchainManager(ThreadSafeManager):
    """Python wrapper for the C++ BlockchainManager class with full thread safety."""
    
    def __init__(self):
        """Initialize the Blockchain Manager."""
        super().__init__("BlockchainManager")
        self._manager = satox_cpp.BlockchainManager()
        self._block_cache = {}
        self._transaction_cache = {}
        self._network_cache = {}
        self._mempool_cache = {}
    
    def initialize(self) -> bool:
        """Initialize the blockchain manager.
        
        Returns:
            bool: True if initialization was successful, False otherwise.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_lock():
                if self._state == ThreadState.INITIALIZED:
                    return True
                
                try:
                    self._set_state(ThreadState.INITIALIZING)
                    success = self._manager.initialize()
                    if success:
                        self._set_state(ThreadState.INITIALIZED)
                        self._clear_error()
                        self.emit_event("initialized")
                        return True
                    else:
                        self._set_error("Failed to initialize blockchain manager")
                        return False
                except Exception as e:
                    self._set_error(f"Initialization error: {str(e)}")
                    return False
        finally:
            self._end_operation(operation_id)
    
    def shutdown(self) -> bool:
        """Shutdown the blockchain manager.
        
        Returns:
            bool: True if shutdown was successful, False otherwise.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_lock():
                if self._state == ThreadState.SHUTDOWN:
                    return True
                
                try:
                    self._set_state(ThreadState.SHUTTING_DOWN)
                    success = self._manager.shutdown()
                    if success:
                        self._set_state(ThreadState.SHUTDOWN)
                        self._clear_error()
                        self.emit_event("shutdown")
                        # Clear all caches
                        self._clear_all_caches()
                        return True
                    else:
                        self._set_error("Failed to shutdown blockchain manager")
                        return False
                except Exception as e:
                    self._set_error(f"Shutdown error: {str(e)}")
                    return False
        finally:
            self._end_operation(operation_id)
    
    def get_current_height(self) -> int:
        """Get the current blockchain height.
        
        Returns:
            int: Current block height
            
        Raises:
            RuntimeError: If the manager is not initialized.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_read_lock():
                if not self._ensure_initialized():
                    raise RuntimeError(self.get_last_error())
                
                try:
                    height = self._manager.getCurrentHeight()
                    self.emit_event("height_queried", height)
                    return height
                except Exception as e:
                    self._set_error(f"Failed to get current height: {str(e)}")
                    raise RuntimeError(self.get_last_error())
        finally:
            self._end_operation(operation_id)
    
    def get_balance(self, address: str) -> float:
        """Get the balance for an address.
        
        Args:
            address (str): The address to get balance for
            
        Returns:
            float: The balance
            
        Raises:
            RuntimeError: If the manager is not initialized.
            ValueError: If the address is invalid.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_read_lock():
                if not self._ensure_initialized():
                    raise RuntimeError(self.get_last_error())
                
                if not address:
                    raise ValueError("Invalid address")
                
                # Check cache first
                cache_key = f"balance_{address}"
                cached_balance = self.get_cache(cache_key)
                if cached_balance is not None:
                    return cached_balance
                
                try:
                    balance = self._manager.getBalance(address)
                    # Cache balance for 30 seconds
                    self.set_cache(cache_key, balance, ttl=30.0)
                    self.emit_event("balance_queried", address, balance)
                    return balance
                except Exception as e:
                    self._set_error(f"Failed to get balance: {str(e)}")
                    raise RuntimeError(self.get_last_error())
        finally:
            self._end_operation(operation_id)
    
    def get_statistics(self) -> Dict[str, Any]:
        """Get blockchain statistics.
        
        Returns:
            Dict[str, Any]: Blockchain statistics
            
        Raises:
            RuntimeError: If the manager is not initialized.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_read_lock():
                if not self._ensure_initialized():
                    raise RuntimeError(self.get_last_error())
                
                # Check cache first
                cache_key = "blockchain_statistics"
                cached_stats = self.get_cache(cache_key)
                if cached_stats is not None:
                    return cached_stats
                
                try:
                    stats = self._manager.getStatistics()
                    # Cache statistics for 60 seconds
                    self.set_cache(cache_key, stats, ttl=60.0)
                    self.emit_event("statistics_queried", stats)
                    return stats
                except Exception as e:
                    self._set_error(f"Failed to get statistics: {str(e)}")
                    raise RuntimeError(self.get_last_error())
        finally:
            self._end_operation(operation_id)
    
    def get_block_info(self, block_hash: str) -> Dict[str, Any]:
        """Get information about a specific block.
        
        Args:
            block_hash (str): The block hash
            
        Returns:
            Dict[str, Any]: Block information
            
        Raises:
            RuntimeError: If the manager is not initialized.
            ValueError: If the block hash is invalid.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_read_lock():
                if not self._ensure_initialized():
                    raise RuntimeError(self.get_last_error())
                
                if not block_hash:
                    raise ValueError("Invalid block hash")
                
                # Check cache first
                cache_key = f"block_{block_hash}"
                cached_info = self.get_cache(cache_key)
                if cached_info is not None:
                    return cached_info
                
                try:
                    block_info = self._manager.getBlockInfo(block_hash)
                    # Cache the result for 5 minutes
                    self.set_cache(cache_key, block_info, ttl=300.0)
                    self.emit_event("block_info_queried", block_hash, block_info)
                    return block_info
                except Exception as e:
                    self._set_error(f"Failed to get block info: {str(e)}")
                    raise RuntimeError(self.get_last_error())
        finally:
            self._end_operation(operation_id)
    
    def get_transaction_info(self, txid: str) -> Dict[str, Any]:
        """Get information about a specific transaction.
        
        Args:
            txid (str): The transaction ID
            
        Returns:
            Dict[str, Any]: Transaction information
            
        Raises:
            RuntimeError: If the manager is not initialized.
            ValueError: If the transaction ID is invalid.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_read_lock():
                if not self._ensure_initialized():
                    raise RuntimeError(self.get_last_error())
                
                if not txid:
                    raise ValueError("Invalid transaction ID")
                
                # Check cache first
                cache_key = f"tx_{txid}"
                cached_info = self.get_cache(cache_key)
                if cached_info is not None:
                    return cached_info
                
                try:
                    tx_info = self._manager.getTransactionInfo(txid)
                    # Cache the result for 2 minutes
                    self.set_cache(cache_key, tx_info, ttl=120.0)
                    self.emit_event("transaction_info_queried", txid, tx_info)
                    return tx_info
                except Exception as e:
                    self._set_error(f"Failed to get transaction info: {str(e)}")
                    raise RuntimeError(self.get_last_error())
        finally:
            self._end_operation(operation_id)
    
    def get_network_info(self) -> Dict[str, Any]:
        """Get network information.
        
        Returns:
            Dict[str, Any]: Network information
            
        Raises:
            RuntimeError: If the manager is not initialized.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_read_lock():
                if not self._ensure_initialized():
                    raise RuntimeError(self.get_last_error())
                
                # Check cache first
                cache_key = "network_info"
                cached_info = self.get_cache(cache_key)
                if cached_info is not None:
                    return cached_info
                
                try:
                    network_info = self._manager.getNetworkInfo()
                    # Cache for 30 seconds
                    self.set_cache(cache_key, network_info, ttl=30.0)
                    self.emit_event("network_info_queried", network_info)
                    return network_info
                except Exception as e:
                    self._set_error(f"Failed to get network info: {str(e)}")
                    raise RuntimeError(self.get_last_error())
        finally:
            self._end_operation(operation_id)
    
    def validate_address(self, address: str) -> bool:
        """Validate a Satoxcoin address.
        
        Args:
            address (str): The address to validate
            
        Returns:
            bool: True if valid, False otherwise
            
        Raises:
            RuntimeError: If the manager is not initialized.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_read_lock():
                if not self._ensure_initialized():
                    raise RuntimeError(self.get_last_error())
                
                if not address:
                    return False
                
                try:
                    is_valid = self._manager.validateAddress(address)
                    self.emit_event("address_validated", address, is_valid)
                    return is_valid
                except Exception as e:
                    self._set_error(f"Failed to validate address: {str(e)}")
                    raise RuntimeError(self.get_last_error())
        finally:
            self._end_operation(operation_id)
    
    def get_mempool_info(self) -> Dict[str, Any]:
        """Get mempool information.
        
        Returns:
            Dict[str, Any]: Mempool information
            
        Raises:
            RuntimeError: If the manager is not initialized.
        """
        operation_id = self._begin_operation()
        try:
            with self._with_read_lock():
                if not self._ensure_initialized():
                    raise RuntimeError(self.get_last_error())
                
                # Check cache first
                cache_key = "mempool_info"
                cached_info = self.get_cache(cache_key)
                if cached_info is not None:
                    return cached_info
                
                try:
                    mempool_info = self._manager.getMempoolInfo()
                    # Cache for 10 seconds (mempool changes frequently)
                    self.set_cache(cache_key, mempool_info, ttl=10.0)
                    self.emit_event("mempool_info_queried", mempool_info)
                    return mempool_info
                except Exception as e:
                    self._set_error(f"Failed to get mempool info: {str(e)}")
                    raise RuntimeError(self.get_last_error())
        finally:
            self._end_operation(operation_id)
    
    def clear_cache(self) -> None:
        """Clear all caches in a thread-safe manner."""
        operation_id = self._begin_operation()
        try:
            with self._with_lock():
                super().clear_cache()
                self._block_cache.clear()
                self._transaction_cache.clear()
                self._network_cache.clear()
                self._mempool_cache.clear()
                self.emit_event("cache_cleared")
        finally:
            self._end_operation(operation_id)
    
    def get_cache_stats(self) -> Dict[str, Any]:
        """Get cache statistics.
        
        Returns:
            Dict[str, Any]: Cache statistics
        """
        operation_id = self._begin_operation()
        try:
            with self._with_read_lock():
                base_stats = super().get_cache_stats()
                additional_stats = {
                    "block_cache_size": len(self._block_cache),
                    "transaction_cache_size": len(self._transaction_cache),
                    "network_cache_size": len(self._network_cache),
                    "mempool_cache_size": len(self._mempool_cache)
                }
                base_stats.update(additional_stats)
                return base_stats
        finally:
            self._end_operation(operation_id)
    
    def _clear_all_caches(self) -> None:
        """Clear all internal caches."""
        self._block_cache.clear()
        self._transaction_cache.clear()
        self._network_cache.clear()
        self._mempool_cache.clear()
    
    def _dispose_impl(self) -> None:
        """Implementation-specific disposal logic."""
        try:
            if hasattr(self, '_manager'):
                self._manager.shutdown()
        except Exception as e:
            self._set_error(f"Disposal error: {str(e)}")
        finally:
            self._clear_all_caches() 

File: bindings/lua/satox_bindings/blockchain_manager.lua
-- Blockchain Manager for Lua Bindings with Full Thread Safety
local ThreadSafeManager = require("satox_bindings.thread_safe_manager")

local BlockchainManager = {}
BlockchainManager.__index = BlockchainManager

-- Inherit from ThreadSafeManager
setmetatable(BlockchainManager, { __index = ThreadSafeManager })

function BlockchainManager.new()
    local self = ThreadSafeManager.new("BlockchainManager")
    setmetatable(self, BlockchainManager)
    
    -- Initialize blockchain-specific data
    self.blocks = {}
    self.transactions = {}
    self.networkConfig = {
        p2pPort = 60777,
        rpcPort = 7777,
        networkType = "mainnet"
    }
    
    -- Cache for frequently accessed data
    self.blockCache = {}
    self.transactionCache = {}
    self.balanceCache = {}
    self.networkCache = {}
    
    return self
end

function BlockchainManager:initialize()
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return false
    end
    
    self:set_state("initializing")
    
    -- Initialize blockchain connection
    local success = self:initialize_blockchain()
    if success then
        self:set_state("initialized")
        self:clear_error()
        self:emit_event("initialized")
        self:end_operation(operation_id)
        return true
    else
        self:set_error("Failed to initialize blockchain manager")
        self:end_operation(operation_id)
        return false
    end
end

function BlockchainManager:shutdown()
    local operation_id = self:begin_operation()
    
    if self:get_state() == "shutdown" then
        self:end_operation(operation_id)
        return true
    end
    
    self:set_state("shutting_down")
    
    -- Shutdown blockchain connection
    local success = self:shutdown_blockchain()
    if success then
        self:set_state("shutdown")
        self:clear_error()
        self:emit_event("shutdown")
        self:clear_all_caches()
        self:end_operation(operation_id)
        return true
    else
        self:set_error("Failed to shutdown blockchain manager")
        self:end_operation(operation_id)
        return false
    end
end

function BlockchainManager:connect()
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return false
    end
    
    -- Check cache first
    local cache_key = "connection_status"
    local cached_status = self:get_cache(cache_key)
    if cached_status and cached_status.connected then
        self:end_operation(operation_id)
        return true
    end
    
    -- Connect to blockchain network
    local success = self:connect_to_network()
    if success then
        local status = { connected = true, timestamp = os.time() }
        self:set_cache(cache_key, status, 30.0)
        self:emit_event("connected", status)
        self:end_operation(operation_id)
        return true
    else
        self:set_error("Failed to connect to blockchain network")
        self:end_operation(operation_id)
        return false
    end
end

function BlockchainManager:disconnect()
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return false
    end
    
    -- Disconnect from blockchain network
    local success = self:disconnect_from_network()
    if success then
        self:remove_cache("connection_status")
        self:emit_event("disconnected")
        self:end_operation(operation_id)
        return true
    else
        self:set_error("Failed to disconnect from blockchain network")
        self:end_operation(operation_id)
        return false
    end
end

function BlockchainManager:is_connected()
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return false
    end
    
    -- Check cache first
    local cache_key = "connection_status"
    local cached_status = self:get_cache(cache_key)
    if cached_status then
        self:end_operation(operation_id)
        return cached_status.connected
    end
    
    -- Check actual connection status
    local connected = self:check_connection_status()
    local status = { connected = connected, timestamp = os.time() }
    self:set_cache(cache_key, status, 30.0)
    
    self:end_operation(operation_id)
    return connected
end

function BlockchainManager:get_block_height()
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return nil
    end
    
    -- Check cache first
    local cache_key = "block_height"
    local cached_height = self:get_cache(cache_key)
    if cached_height then
        self:end_operation(operation_id)
        return cached_height
    end
    
    -- Get current block height
    local height = self:fetch_block_height()
    if height then
        self:set_cache(cache_key, height, 60.0)
        self:emit_event("height_queried", height)
        self:end_operation(operation_id)
        return height
    else
        self:set_error("Failed to get block height")
        self:end_operation(operation_id)
        return nil
    end
end

function BlockchainManager:get_block(block_hash)
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return nil
    end
    
    if not block_hash then
        self:set_error("Invalid block hash")
        self:end_operation(operation_id)
        return nil
    end
    
    -- Check cache first
    local cache_key = "block_" .. block_hash
    local cached_block = self:get_cache(cache_key)
    if cached_block then
        self:end_operation(operation_id)
        return cached_block
    end
    
    -- Fetch block information
    local block_info = self:fetch_block_info(block_hash)
    if block_info then
        self:set_cache(cache_key, block_info, 300.0) -- Cache for 5 minutes
        self:emit_event("block_info_queried", { hash = block_hash, info = block_info })
        self:end_operation(operation_id)
        return block_info
    else
        self:set_error("Failed to get block info")
        self:end_operation(operation_id)
        return nil
    end
end

function BlockchainManager:get_transaction(txid)
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return nil
    end
    
    if not txid then
        self:set_error("Invalid transaction ID")
        self:end_operation(operation_id)
        return nil
    end
    
    -- Check cache first
    local cache_key = "tx_" .. txid
    local cached_tx = self:get_cache(cache_key)
    if cached_tx then
        self:end_operation(operation_id)
        return cached_tx
    end
    
    -- Fetch transaction information
    local tx_info = self:fetch_transaction_info(txid)
    if tx_info then
        self:set_cache(cache_key, tx_info, 120.0) -- Cache for 2 minutes
        self:emit_event("transaction_info_queried", { txid = txid, info = tx_info })
        self:end_operation(operation_id)
        return tx_info
    else
        self:set_error("Failed to get transaction info")
        self:end_operation(operation_id)
        return nil
    end
end

function BlockchainManager:send_transaction(transaction)
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return nil
    end
    
    if not transaction then
        self:set_error("Invalid transaction")
        self:end_operation(operation_id)
        return nil
    end
    
    -- Validate transaction
    local is_valid = self:validate_transaction(transaction)
    if not is_valid then
        self:set_error("Transaction validation failed")
        self:end_operation(operation_id)
        return nil
    end
    
    -- Send transaction
    local txid = self:broadcast_transaction(transaction)
    if txid then
        self:emit_event("transaction_sent", { txid = txid, transaction = transaction })
        self:end_operation(operation_id)
        return txid
    else
        self:set_error("Failed to send transaction")
        self:end_operation(operation_id)
        return nil
    end
end

function BlockchainManager:get_balance(address)
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return nil
    end
    
    if not address then
        self:set_error("Invalid address")
        self:end_operation(operation_id)
        return nil
    end
    
    -- Check cache first
    local cache_key = "balance_" .. address
    local cached_balance = self:get_cache(cache_key)
    if cached_balance then
        self:end_operation(operation_id)
        return cached_balance
    end
    
    -- Fetch balance
    local balance = self:fetch_balance(address)
    if balance then
        self:set_cache(cache_key, balance, 30.0) -- Cache for 30 seconds
        self:emit_event("balance_queried", { address = address, balance = balance })
        self:end_operation(operation_id)
        return balance
    else
        self:set_error("Failed to get balance")
        self:end_operation(operation_id)
        return nil
    end
end

function BlockchainManager:validate_address(address)
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return false
    end
    
    if not address then
        self:end_operation(operation_id)
        return false
    end
    
    -- Validate address
    local is_valid = self:validate_satox_address(address)
    self:emit_event("address_validated", { address = address, valid = is_valid })
    
    self:end_operation(operation_id)
    return is_valid
end

function BlockchainManager:get_network_status()
    local operation_id = self:begin_operation()
    
    if not self:ensure_initialized() then
        self:end_operation(operation_id)
        return nil
    end
    
    -- Check cache first
    local cache_key = "network_status"
    local cached_status = self:get_cache(cache_key)
    if cached_status then
        self:end_operation(operation_id)
        return cached_status
    end
    
    -- Get network status
    local status = self:fetch_network_status()
    if status then
        self:set_cache(cache_key, status, 60.0) -- Cache for 1 minute
        self:emit_event("network_status_queried", status)
        self:end_operation(operation_id)
        return status
    else
        self:set_error("Failed to get network status")
        self:end_operation(operation_id)
        return nil
    end
end

function BlockchainManager:clear_cache()
    local operation_id = self:begin_operation()
    
    -- Clear all caches
    self:clear_all_caches()
    self:emit_event("cache_cleared")
    
    self:end_operation(operation_id)
end

-- Private helper methods (stubs for actual implementation)
function BlockchainManager:initialize_blockchain()
    -- TODO: Implement actual blockchain initialization
    return true
end

function BlockchainManager:shutdown_blockchain()
    -- TODO: Implement actual blockchain shutdown
    return true
end

function BlockchainManager:connect_to_network()
    -- TODO: Implement actual network connection
    return true
end

function BlockchainManager:disconnect_from_network()
    -- TODO: Implement actual network disconnection
    return true
end

function BlockchainManager:check_connection_status()
    -- TODO: Implement actual connection status check
    return true
end

function BlockchainManager:fetch_block_height()
    -- TODO: Implement actual block height fetching
    return #self.blocks
end

function BlockchainManager:fetch_block_info(block_hash)
    -- TODO: Implement actual block info fetching
    return { hash = block_hash, height = 1, timestamp = os.time() }
end

function BlockchainManager:fetch_transaction_info(txid)
    -- TODO: Implement actual transaction info fetching
    return { txid = txid, block_height = 1, confirmations = 1 }
end

function BlockchainManager:validate_transaction(transaction)
    -- TODO: Implement actual transaction validation
    return true
end

function BlockchainManager:broadcast_transaction(transaction)
    -- TODO: Implement actual transaction broadcasting
    local txid = "tx_" .. #self.transactions + 1
    self.transactions[txid] = transaction
    return txid
end

function BlockchainManager:fetch_balance(address)
    -- TODO: Implement actual balance fetching
    return 0.0
end

function BlockchainManager:validate_satox_address(address)
    -- TODO: Implement actual address validation
    return string.len(address) > 0
end

function BlockchainManager:fetch_network_status()
    -- TODO: Implement actual network status fetching
    return {
        connected = true,
        peers = 10,
        block_height = #self.blocks,
        difficulty = 1.0
    }
end

function BlockchainManager:clear_all_caches()
    self.blockCache = {}
    self.transactionCache = {}
    self.balanceCache = {}
    self.networkCache = {}
end

return BlockchainManager 